eclipse-theia/theia
# process/task/terminal: refactor escaping/quoting

#### What it does

Add utility functions in `@theia/process/lib/common` to escape common
shells' arguments. Refactored terminal processes to not handle shell
escaping anymore, it is the caller's responsability to provide the
escaped spawn options.

Escaping is now done for DAP's `runInTerminal` requests.

Most of the design is taken from VS Code, with minor tweaks.

Tried to test as much as possible since escaping rules can get complicated in corner cases.

Fixes https://github.com/eclipse-theia/theia/issues/6684

#### How to test

1. Build and then run the test suites (`@theia/process`, `@theia/task`, and `@theia/debug`).
2. Test debugging a Python file:
    1. Download [ms-python.vsix](https://github.com/Microsoft/vscode-python/releases/download/2019.5.18875/ms-python-release.vsix)
    2. Setup debug config to run opened Python files.
    3. Try to debug a Python file with space in its filename, it should work with this patch but not without.
3. Test different `shell` tasks, using quoted strings (see [VS Code Custom Tasks](https://code.visualstudio.com/docs/editor/tasks#_custom-tasks)).

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
